src/script/tari_script.rs

```diff
@@ -138,41 +196,193 @@ impl TariScript {
+            Sub => TariScript::handle_op_sub(stack),
+            Equal => match TariScript::handle_equal(stack)? {
+                true => stack.push(Number(1)),
+                false => stack.push(Number(0)),
             },
-            Opcode::EqualVerify => match TariScript::handle_equal(stack)? {
-                true => stack.push(Number(0)),
+            EqualVerify => match TariScript::handle_equal(stack)? {
+                true => stack.push(Number(1)),
                 false => Err(ScriptError::VerifyFailed),
             },
-            Opcode::CheckSig => match self.check_sig(stack)? {
-                true => stack.push(Number(0)),
-                false => stack.push(Number(1)),
+            Or(n) => TariScript::handle_or(stack, *n),
+            OrVerify(n) => TariScript::handle_or_verify(stack, *n),
+            HashBlake256 => TariScript::handle_hash::<Blake256>(stack),
+            HashSha256 => TariScript::handle_hash::<Sha256>(stack),
+            HashSha3 => TariScript::handle_hash::<Sha3>(stack),
+            CheckSig(msg) => match self.check_sig(stack, *msg.deref())? {
+                true => stack.push(Number(1)),
+                false => stack.push(Number(0)),
             },
-            Opcode::CheckSigVerify => match self.check_sig(stack)? {
-                true => stack.push(Number(0)),
-                false => Err(ScriptError::InvalidSignature),
+            CheckSigVerify(msg) => match self.check_sig(stack, *msg.deref())? {
+                true => Ok(()),
+                false => Err(ScriptError::VerifyFailed),
             },
-            Opcode::RevRot => stack.push_down(2),
-            Opcode::PushHash(h) => stack.push(Hash(*h.clone())),
-            Opcode::HashBlake256 => TariScript::handle_hash::<Blake256>(stack),
-            Opcode::PushZero => stack.push(Number(0)),
+            Return => Err(ScriptError::Return),
+            IfThen => self.handle_if_then(stack),
+            Else => self.handle_else(),
+            EndIf => self.handle_end_if(),
+        }
+    }
+
+    fn handle_check_height_verify(height: u64, block_height: u64) -> Result<(), ScriptError> {
+        if block_height >= height {
+            Ok(())
+        } else {
+            Err(ScriptError::VerifyFailed)
+        }
+    }
+
+    fn handle_check_height(stack: &mut ExecutionStack, height: u64, block_height: u64) -> Result<(), ScriptError> {
+        let height = i64::try_from(height)?;
+        let block_height = i64::try_from(block_height)?;
+        let item = StackItem::Number(block_height - height);
+
+        stack.push(item)
+    }
+
+    fn handle_compare_height_verify(stack: &mut ExecutionStack, block_height: u64) -> Result<(), ScriptError> {
+        let item = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let target_height = if let StackItem::Number(n) = item {
+            Ok(n)
+        } else {
+            Err(ScriptError::InvalidInput)
+        }?;
```

Comment:
Not sure I prefer this, but too bothered I'll update it. Does it pass clippy?

Reply:
I don't think clippy complains about the `if {Ok()} else{Err()}?`. Not a big deal at all, and I know I'm nitting. It's easy to miss the `?` at the end of the `if` when reading.  Generally not fond of constructing a `Result` just so that `?` can be used when a `match` is clearer.

```rust
// De-sugared `?`
let target_height = match (if let StackItem::Number(n) = item {
            Ok(n)
        } else {
            Err(ScriptError::InvalidInput)
        }) {
Ok(n) => n,
Err(err) => {return Err(err.into()); }
};
```